pkg/testutils/kubernetes/builders/client.go

```diff
@@ -0,0 +1,212 @@
+package builders
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+const (
+	// ObjectEventAll subscribe to all object events
+	ObjectEventAll ObjectEvent = "ALL"
+	// ObjectEventAdded subscribe to object creation events
+	ObjectEventAdded ObjectEvent = "ADDED"
+	// ObjectEventDeleted subscribe to object delete events
+	ObjectEventDeleted ObjectEvent = "DELETED"
+	// ObjectEventModified subscribe to object update events
+	ObjectEventModified ObjectEvent = "MODIFIED"
+)
+
+// ObjectEvent defines an event in an object
+type ObjectEvent string
+
+// PodObserver is a function that receives notifications of events on an Pod
+// and can update it by returning a non-nil value. In addition, the PodObserver
+// returns a boolean value indicating if it wants to keep receiving events or not.
+//
+// Note: PodObserver that subscribe to update events should implement a mechanism for
+// avoiding an update loop. They can for instance check the object is in a particular state
+// before updating. Additionally, they can unsubscribe from further updates.
+type PodObserver func(ObjectEvent, *corev1.Pod) (*corev1.Pod, bool, error)
```

Comment:
I find the signature of the observer function a bit confusing:
- When I register an observer I already supply an `ObjectEvent`. Does the function need to receive it again when it is called?
- Returning nil as an indication of not doing anything is a valid option, but I think it is more explicit and readable if the observer function actively chooses to not do anything. This could be done, for example, by having the observer function receive as an argument a function to edit the pod, which then the observer decides whether to call or not.
- Returning both a boolean and an error seems a bit redundant, specially since both do the same thing: Breaking the observer loop. I'd suggest to return a special error instead and use `error.Is` in the loop to decide whether the error should be logged. This is a common pattern in the stdlib, e.g. [`WalkDirFunc`](https://pkg.go.dev/io/fs#WalkDirFunc) or [`SplitFunc`](https://pkg.go.dev/bufio#SplitFunc).

To exemplify the suggestions below, the signature could be:

```
var ErrStopWatching = errors.New("stop watching")

// PodMutatorFunc is a function that is called when a pod is added, deleted,
// or modified, and that can react to this event by mutating the pod by calling
// `mutate` with a the modified pod as an argument.
// PodMutatorFunc may return ErrStopWatching if they do not wish to continue being
// called for any other pod event.
type PodMutatorFunc func(pod *corev1.Pod, mutate func(*corev1.Pod) error) error
```

Reply:
> When I register an observer I already supply an ObjectEvent. Does the function need to receive it again when it is called?
Yes, because there is the option to subscribe to all. 

> by having the observer function receive as an argument a function to edit the pod

I find this option overcomplicated and, as you said,  "returning nil is a valid option". I prefer to keep it simple. if we later find there are use cases that are not well covered by just retuning nil, we can consider this approach. 

> Returning both a boolean and an error seems a bit redundant,

No, they are not redundant because the boolean indicates the observer doesn't;t want to continue observing and the error means there was an error processing the event, two very different situations

Reply:
I don't see how passing a function is overcomplicated, neither I agree that checking error types is complex. The standard library is doing (thus, by example, encouraging) this pattern in multiple instances.

I wouldn't like to block the PR over this though, specially being a very internal API. Let's keep it as it is if you think it's cleaner.

Reply:
> I don't see how passing a function is overcomplicated

Because there's a simpler option that is retuning a value. 

>  neither I agree that checking error types is complex

it is not complex if the error is an error, but it is when errors are abused to signal non-error situations.